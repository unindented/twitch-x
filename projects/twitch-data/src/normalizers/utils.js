export function valueOrNull (value) {
  return value || null
}
