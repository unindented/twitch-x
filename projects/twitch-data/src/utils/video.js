import {stringify} from 'qs'
import {USHER_API_URL} from '../constants'

export function getPlaylistUrl ({stream, sig, token}) {
  const query = {
    sig,
    token,
    allow_audio_only: true,
    allow_source: true,
    player: 'twitchweb',
    type: 'any',
    p: Math.floor(Math.random() * 1e7)
  }
  return `${USHER_API_URL}/channel/hls/${stream}.m3u8?${stringify(query)}`
}

export function extractFeedsFromPlaylist (input) {
  const regexp = /#EXT-X-MEDIA:.*NAME="([^"]+)".*\n#EXT-X-STREAM-INF:.*\n(.*)/g
  const result = []

  let match = regexp.exec(input)
  while (match != null) {
    result.push({
      name: match[1],
      url: match[2]
    })
    match = regexp.exec(input)
  }

  return result
}
