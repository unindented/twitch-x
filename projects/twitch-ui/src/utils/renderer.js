import React from 'react'
import renderer from 'react-test-renderer'
import {ThemeProvider} from 'styled-components'
import themes from '../themes'

export function renderWithTheme (node, theme = 'default') {
  return renderer.create(
    <ThemeProvider theme={themes[theme]}>
      {node}
    </ThemeProvider>
  )
}
