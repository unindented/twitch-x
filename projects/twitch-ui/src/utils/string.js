export function supplant (str, obj) {
  return str.replace(/\{\s*([^{}]+?)\s*\}/g, (str, key) => obj[key])
}
