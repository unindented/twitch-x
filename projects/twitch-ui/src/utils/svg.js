export function rectangle ({width, height, fill}) {
  return `<svg viewBox="0 0 ${width} ${height}" xmlns="http://www.w3.org/2000/svg"><rect fill="${fill}" width="100%" height="100%" /></svg>`
}
