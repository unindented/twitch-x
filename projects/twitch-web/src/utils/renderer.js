import ShallowRenderer from 'react-test-renderer/shallow'

export function render (node, theme = 'default') {
  const renderer = new ShallowRenderer()
  renderer.render(node)
  return renderer.getRenderOutput()
}
