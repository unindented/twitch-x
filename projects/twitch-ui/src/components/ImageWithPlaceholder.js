import React, {PureComponent} from 'react'
import PropTypes from 'prop-types'
import styled, {withTheme} from 'styled-components'
import {supplant} from '../utils/string'
import {rectangle} from '../utils/svg'

const Image = styled.img`
  display: ${props => props.hidden ? 'none' : 'block'};
  width: 100%;
  height: auto;
`

class ImageWithPlaceholder extends PureComponent {
  constructor (props) {
    super(props)

    this.state = {loaded: false}

    this.handleLoad = this.handleLoad.bind(this)
  }

  handleLoad () {
    this.setState({loaded: true})
  }

  render () {
    const {loaded} = this.state
    const {alt, src, width, height, theme} = this.props
    const opts = {width, height, fill: theme.colors.primary}

    return (
      <div>
        <Image
          alt={alt}
          src={`data:image/svg+xml;utf8,${rectangle(opts)}`}
          hidden={loaded}
        />
        <Image
          alt={alt}
          src={supplant(src, opts)}
          hidden={!loaded}
          onLoad={this.handleLoad}
        />
      </div>
    )
  }
}

ImageWithPlaceholder.propTypes = {
  alt: PropTypes.string,
  src: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  theme: PropTypes.any.isRequired
}

ImageWithPlaceholder.defaultProps = {
  alt: ''
}

export default withTheme(ImageWithPlaceholder)
