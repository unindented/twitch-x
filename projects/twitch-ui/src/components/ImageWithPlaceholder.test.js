import React from 'react'
import {renderWithTheme} from '../utils/renderer'
import ImageWithPlaceholder from './ImageWithPlaceholder'

describe('ImageWithPlaceholder', () => {
  let tree

  beforeEach(() => {
    tree = renderWithTheme(
      <ImageWithPlaceholder
        alt='Some image'
        src='//lorempixel.com/{width}/{height}/'
        width={272}
        height={380}
      />
    )
  })

  it('renders', () => {
    expect(tree).toMatchSnapshot()
  })
})
