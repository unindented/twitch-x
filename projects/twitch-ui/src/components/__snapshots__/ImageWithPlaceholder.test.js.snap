// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageWithPlaceholder renders 1`] = `
.c0 {
  display: block;
  width: 100%;
  height: auto;
}

.c1 {
  display: none;
  width: 100%;
  height: auto;
}

<div>
  <img
    alt="Some image"
    className="c0"
    hidden={false}
    src="data:image/svg+xml;utf8,<svg viewBox=\\"0 0 272 380\\" xmlns=\\"http://www.w3.org/2000/svg\\"><rect fill=\\"#6441a4\\" width=\\"100%\\" height=\\"100%\\" /></svg>"
  />
  <img
    alt="Some image"
    className="c1"
    hidden={true}
    onLoad={[Function]}
    src="//lorempixel.com/272/380/"
  />
</div>
`;
